pysparkling/utils.py

```diff
@@ -329,5 +329,39 @@ def merge_rows(left, right, on=_sentinel):
     ))
 
 
+def strhash(string):
+    """
+    Old python hash function as described in PEP 456, excluding prefix, suffix and mask.
+
+    :param string: string to hash
+    :return: hash
+    """
+    if string == "":
+        return 0
+
+    x = ord(string[0]) << 7
+    for c in string[1:]:
+        x = ((1000003 * x) ^ ord(c)) & (1 << 32)
+    x = (x ^ len(string))
+    return x
+
+
+def tuplehash(tuple):
+    """
+    Old python hash function for tuples.
+
+    :param tuple: tuple to hash
+    :return: hash
+    """
+    if isinstance(tuple, int):
+        return tuple
+    mult = 1000003
+    x = 0x345678
+    for index, item in enumerate(tuple):
+        x = ((x ^ hash(item)) * mult) & (1<<32)
+        mult += (82520 + (len(tuple)-index)*2)
+    return x + 97531
+
+
 def portable_hash(item):
-    return hash(item) & 0xffffffff
+    return strhash(item) & 0xffffffff
```

Comment:
Thanks for the improvement,

Do you believe we could use the code of `portable_hash` as defined in https://github.com/apache/spark/blob/master/python/pyspark/rdd.py#L80-L108 and add a condition if item is a string? Currently `item` can be something else than a string, e.g. in `rdd.partitionBy(num_partitions, portable_hash)` as used in `hash_partition_and_sort`

Reply:
This way we end up with a `portable_hash` which is more portable than pyspark's one :)

Reply:
I think we can use the function from pyspark with minor changes. I'll see what I can do.